sdk/python/feast/registry.py

```diff
@@ -0,0 +1,123 @@
+            if entity_proto.spec.name == name:
+                return Entity.from_proto(entity_proto)
+        raise Exception(f"Entity {name} does not exist")
+
+    def apply_feature_table(self, feature_table: FeatureTable):
+        feature_table.is_valid()
+        feature_table_proto = feature_table.to_proto()
+        def updater(registry_proto: RegistryProto):
+            for idx, existing_feature_table_proto in enumerate(registry_proto.feature_tables):
+                if existing_feature_table_proto.spec.name == feature_table_proto.spec.name:
+                    registry_proto.feature_tables[idx] = feature_table_proto
+                    return registry_proto
+            registry_proto.feature_tables.append(feature_table_proto)
+            return registry_proto
+        self._update_registry(updater)
+        return
+
+    def list_feature_tables(self):
+        registry_proto = self._get_registry()
+        feature_tables = []
+        for feature_table_proto in registry_proto.feature_tables:
+            feature_tables.append(FeatureTable.from_proto(feature_table_proto))
+        return feature_tables
+
+    def get_feature_table(self, name: str):
+        registry_proto = self._get_registry()
+        for feature_table_proto in registry_proto.feature_tables:
+            if feature_table_proto.spec.name == name:
+                return FeatureTable.from_proto(feature_table_proto)
+        raise Exception(f"Feature table {name} does not exist")
+
+    def delete_feature_table(self, name: str):
+        def updater(registry_proto: RegistryProto):
+            for idx, existing_feature_table_proto in enumerate(registry_proto.feature_tables):
+                if existing_feature_table_proto.spec.name == feature_table_proto.spec.name:
+                    del registry_proto.feature_tables[idx]
+                    return registry_proto
+            raise Exception(f"Feature table {name} does not exist")
+        self._update_registry(updater)
+        return
+
+    @abstractmethod
+    def _get_registry(self):
+        pass
+
+    @abstractmethod
+    def _update_registry(self, updater: Callable[[RegistryProto], RegistryProto]):
+        pass
+
+
+class LocalRegistry(Registry):
+    def __init__(self, filepath: str):
+        self._filepath = Path(filepath)
+        if not self._filepath.exists():
+            registry_proto = RegistryProto()
+            registry_proto.registry_schema_version = REGISTRY_SCHEMA_VERSION
+            registry_proto.version = str(uuid.uuid4())
+            registry_proto.last_updated.FromDatetime(datetime.utcnow())
+            self._filepath.write_bytes(registry_proto.SerializeToString())
+        return
+
+    def _get_registry(self):
+        registry_proto = RegistryProto()
+        registry_proto.ParseFromString(self._filepath.read_bytes())
+        return registry_proto
+
+    def _update_registry(self, updater: Callable[[RegistryProto], RegistryProto]):
```

Comment:
I need to come back to this code, but I find the classes and methods here surprisingly complex for what is essentially CRUD.

Is there a way that we can simplify the code and make it more readable/explicit by not passing methods around? I'd be quite happy with a simple base class that indicates to implementers what is expected (acts almost like an interface), with an implementation that contains the majority of the logic. 

MLflow for inspiration: https://github.com/mlflow/mlflow/blob/master/mlflow/store/tracking/abstract_store.py

Reply:
As someone who propsed this interface in the first place; the idea with this was to make it possible for the implementation to use optimistic locking for read-update-write, and to make the interface as minimal as possible. This is a low level class that would have a nicer higher level wrapper (like the Registry above)

Reply:
> This is a low level class that would have a nicer higher level wrapper 

Which class are you referring to here, the `LocalRegistry`?

Reply:
Yes (i wouldn't have it inherit from Registry tho)